Optimize follow logs implementation

MadhavJivrajani opened the pull request:
The CPU usage is nearly 0% for the `follow` logs implementation
Implemented using the `subprocess` module.

![Screenshot from 2021-01-29 21-24-01](https://user-images.githubusercontent.com/12381034/106297432-b4920c80-6278-11eb-9a5b-9c3e0d5460ab.png)
![Screenshot from 2021-01-29 21-24-49](https://user-images.githubusercontent.com/12381034/106297441-b8259380-6278-11eb-9b5e-0b8ff757b523.png)

- [x] ran the code through `black`
Fixes #7

Samyak2 commented on heiko/cli.py:
```diff
@@ -119,28 +120,20 @@ def make_parser():
-    line = ""
-    i = 0
+
+    f = subprocess.Popen(
+        ["tail", "-F", file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
     while True:
```
```suggestion
    f = subprocess.Popen(
        ["tail", "-f", file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE
    )
    for line in f.stdout:
```

Samyak2 replied:
This (along with the next suggestion - GitHub wouldn't let me suggest changes to the whole thing at once) should fix the issue I described.

MadhavJivrajani replied:
I had tried this locally, it doesn't print logs as and when appended. Only if you stop the command will it print the logs, I'm not sure why that happens yet.

MadhavJivrajani replied:
We can do 
```py
f = subprocess.Popen(
        ["tail", "-f", file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
)
for line in iter(f.stdout.readline, ""):
        yield line

Samyak2 replied:
> I had tried this locally, it doesn't print logs as and when appended. Only if you stop the command will it print the logs, I'm not sure why that happens yet.

Oof. I only checked if it worked when an invalid file was given.




> We can do this, this works:
> 
> ```python
> f = subprocess.Popen(
>         ["tail", "-f", file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
> )
> for line in f.stdout:
>         yield line
> ```

That works! But maybe use `text=True` instead of `universal_newlines=True`. From the [docs](https://docs.python.org/3/library/subprocess.html#popen-constructor):
> The universal_newlines argument is equivalent to text and is provided for backwards compatibility.

---

> Also, what was the issue with
> 
> ```python
> while True:
>      yield f.stdout.readline().decode()
> ```

The issue was that, when the file didn't exist and `stdout` was empty, it would be stuck in an infinite loop there (also it probably would have entered an infinite loop after the file ends, not sure though).

(in hindsight, I should have explained the changes earlier xD)

MadhavJivrajani replied:
Hey, I found another way that might be neater:
```py
subprocess.run(["tail", "-f", file_path], stdout=sys.stdout, stderr=sys.stderr, check=True)
```
> The issue was that, when the file didn't exist and stdout was empty, it would be stuck in an infinite loop there (also it probably would have entered an infinite loop after the file ends, not sure though).

Oh right, makes sense

Samyak2 replied:
> Hey, I found another way that might be neater, the CPU usage is the same as the current commit, constant 0%
> 
> ```python
> subprocess.run(["tail", "-f", file_path], stdout=sys.stdout, stderr=sys.stderr, check=True)
> ```

Oh that's great, it works well. One issue though - when the file doesn't exist, the error is a bit verbose and doesn't indicate the issue to the user directly (the issue being that the `name` is wrong). A try-except around it with an error message printed on `subprocess.CalledProcessError` would be nice.

MadhavJivrajani replied:
Hmm.
I have one question though. The file not existing error will be raised if the `--name` flag has an invalid value set. Shouldn't that be handled before the execution of following logs in executed? That's done while handling the `init` command in `cli.py`. The reason why handling this explicitly beforehand might be better is because the error message can clearly say that the `--name` value does not exist or something equivalent?

Samyak2 replied:
That's a great point. The `name` should be checked before any of the commands are executed. IIRC the only way to check if the name is correct is to see if the log file exists (`heiko_<name>.out`). Since `start`, `stop` and `restart` check for the name using the PID file and `init` does not need the name to exist, the only other sub-command that needs `name` to exist is `logs`.

We should add a check for the log file right at the start of the `elif args.command == "logs":` block and print an error if it doesn't exist. Let me know I missed something (or if I missed the point xD). If you wish, you can open a separate issue for this and I can merge this PR with the `subprocess.run` change you found earlier.

Samyak2 commented on heiko/cli.py:
```diff
@@ -119,28 +120,20 @@ def make_parser():
-                i = 0
-        else:
-            time.sleep(0.1)
+        yield f.stdout.readline().decode()
```
```suggestion
        yield line
```

MadhavJivrajani replied:
I think `decode()` might be needed because the items that `f.stdout` produces will be a byte strings.

MadhavJivrajani replied:
If the above approach that I mentioned is used, then `decode()` isn't needed

Samyak2 replied:
Right

Samyak2 reviewed (changes requested):
CPU Usage is at a constant 0.0. This is nice!

There is one issue though. Since you're using `-F` (as opposed to `-f`), when the user provides a name that does not exist, `tail` keeps trying to open the file and the command gets stuck there without any error.

MadhavJivrajani commented:
Hey, 
I've added the check and I switched to `check_call()` instead of `run()` since it raises the `subprocess.CalledProcessError` for any non-zero return codes.

MadhavJivrajani commented:
(also, if this looks good, don't merge yet, I wanted to checkout one other thing)

Samyak2 commented on heiko/cli.py:
```diff
@@ -119,28 +120,28 @@ def make_parser():
+    """Checks if Heiko Daemon name exists or not
 
-    :param file: file handle to read
-    :type file: file handle
-    :yield: a newline terminated line or a string of 5 characters
-    :rtype: Iterator[str]
+    Existence of daemon name is done by checking if the 
+    corresponding log file exists.
+
+    :param path_to_log: path to where log file should exist
+    :type path_to_log: str
+    :rtype: bool
     """
```
Since this function isn't really limited to checking the log file and it can be used for any path, you could make this a `file_exists` function that works for any path. The current name and doc would make sense if just the name was passed to it, but as it stands rn the function requires the user (of the function) to get the path to the log file beforehand.

MadhavJivrajani replied:
Makes sense, will do

Samyak2 replied:
Can you also change the doc string to reflect the new name?

MadhavJivrajani replied:
Will do

Samyak2 reviewed (commented):
LGTM!

Just one thing

Samyak2 reviewed (approved):
LGTM. Thank you!
